taskiq/middlewares/smart_retry_middleware.py

```diff
@@ -0,0 +1,177 @@
+    """
+
+    def __init__(
+        self,
+        default_retry_count: int = 3,
+        default_retry_label: bool = False,
+        no_result_on_retry: bool = True,
+        default_delay: float = 5,
+        use_jitter: bool = False,
+        use_delay_exponent: bool = False,
+        max_delay_exponent: float = 60,
+        schedule_source: ScheduleSource | None = None,
+    ) -> None:
+        """
+        Initialize retry middleware.
+
+        :param default_retry_count: Default max retries if not specified.
+        :param default_retry_label: Whether to retry tasks by default.
+        :param no_result_on_retry: Replace result with NoResultError on retry.
+        :param default_delay: Delay in seconds before retrying.
+        :param use_jitter: Add random jitter to retry delay.
+        :param use_delay_exponent: Apply exponential backoff to delay.
+        :param max_delay_exponent: Maximum allowed delay when using backoff.
+        :param schedule_source: Schedule source to use for scheduling.
+            If None, the default broker will be used.
+        """
+        super().__init__()
+        self.default_retry_count = default_retry_count
+        self.default_retry_label = default_retry_label
+        self.no_result_on_retry = no_result_on_retry
+        self.default_delay = default_delay
+        self.use_jitter = use_jitter
+        self.use_delay_exponent = use_delay_exponent
+        self.max_delay_exponent = max_delay_exponent
+        self.schedule_source = schedule_source
+
+    def is_retry_on_error(self, message: TaskiqMessage) -> bool:
+        """
+        Check if retry is enabled for this task.
+
+        Looks for `retry_on_error` label, falls back to default.
+
+        :param message: Original task message.
+        :return: True if should retry on error.
+        """
+        retry_on_error = message.labels.get("retry_on_error")
+        if isinstance(retry_on_error, str):
+            retry_on_error = retry_on_error.lower() == "true"
+        if retry_on_error is None:
+            retry_on_error = self.default_retry_label
+        return retry_on_error
+
+    def make_delay(self, message: TaskiqMessage, retries: int) -> float:
+        """
+        Calculate retry delay.
+
+        Includes jitter and exponential backoff if enabled.
+
+        :param message: Task message.
+        :param retries: Current retry count.
+        :return: Delay in seconds.
+        """
+        delay = float(message.labels.get("delay", self.default_delay))
+        if self.use_delay_exponent:
+            delay = min(delay * retries, self.max_delay_exponent)
+
+        if self.use_jitter:
+            delay += random.random()  # noqa: S311
+
+        return delay
+
+    async def on_send(
+        self,
+        kicker: AsyncKicker[Any, Any],
+        message: TaskiqMessage,
+        delay: float,
+    ) -> None:
+        """Execute the task with a delay."""
+        if isinstance(self.schedule_source, ScheduleSource):
```

Comment:
This condition can be simplified a bit: 

```suggestion
        if self.schedule_source is not None:
```